### PR TITLE
Detailed Design 

### PR DESCRIPTION
**Pull request by nainasingh**
For GISTIC2 CNA import

### CODE REVIEW

File: tests/testthat/testImportCNA.GISTIC2.R [hunk 1]
```diff
@@ -0,0 +1,21 @@
+#
+# test importCNA.GISTIC2.R
+#
```
Comment (hyginn) [inline tests/testthat/testImportCNA.GISTIC2.R]: I recommend to follow the structure given in testDummy.R - test handling of erroneous parameters and corrupt input, then check whether sane input gives correct output.

File: R/importCNA.GISTIC2.R [hunk 2]
```diff
@@ -0,0 +1,105 @@
+# importCNA.GISTIC2.R
+#
+#
+# Imports copy number aberration data from Firehose database,
+# produces an rCNA object saved as a RDS file.
+#
+#
+#'@param fFHG A vector of local fully qualified paths of Firehose GISTIC2 CNA file(s).
+#'@param dCNA The local fully qualified path of a directory to store the rCNA RDS file(s).
+#'@param silent Boolean. Default: FALSE. Whether or not to write progress information to console.
```
Comment (hyginn) [inline R/importCNA.GISTIC2.R]: I think the function should operate on one file at a time. The user should supply a valid filename, not merely a directory. Otherwise you would need a large mechanism for creating unique filenames - and that's really not the job of this function.

File: R/importCNA.GISTIC2.R [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+# Rownames: HGNC gene symbols.
+# Column names: Tumor Sample Barcodes
+# Values: CNA as copynumber -2. Floating point number or NA
+#
+# TODO
+# MAKE SURE PATH GIVEN IS VALID AND FILE IS NOT EMPTY/CORRUPT
+# READ
+# GET RID OF INFORMATION NOT NEEDED (FILTER)
+# SAVE AS DATAFRAME AS RDS,  AND STORE IN DIR
+# PRINT STATUS
```
Comment (hyginn) [inline R/importCNA.GISTIC2.R]: You'll need to define what summary information is useful.

File: R/importCNA.GISTIC2.R [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+#
+#
+importCNA.GISTIC2<- function(fFHG, dCNA, silent = FALSE, noLog = FALSE){
+
+    # start log file
```
Comment (hyginn) [inline R/importCNA.GISTIC2.R]: This will need to change - check the importNet.STRING code after the next push.

File: R/importCNA.GISTIC2.R [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        write("Log begin",file="test.log",append=TRUE)
+        write(format(Sys.time(), "%a %b %d %X %Y"),file="test.log",append=TRUE)
+        write(fFHG,file="test.log",append=TRUE)
+        write(dCNA,file="test.log",append=TRUE)
+    }
+
+    # check if file exists
+
+    numFiles <- length(fFHG)
+    for (file in 1:numFiles){
```
Comment (hyginn) [inline R/importCNA.GISTIC2.R]: use the .checkArgs() function for parameter checks (again: cf importNet.STRING())

File: R/importCNA.GISTIC2.R [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+# TODO
+# MAKE SURE PATH GIVEN IS VALID AND FILE IS NOT EMPTY/CORRUPT
+# READ
+# GET RID OF INFORMATION NOT NEEDED (FILTER)
+# SAVE AS DATAFRAME AS RDS,  AND STORE IN DIR
+# PRINT STATUS
+#
+#
+importCNA.GISTIC2<- function(fFHG, dCNA, silent = FALSE, noLog = FALSE){
+
```
Comment (hyginn) [inline R/importCNA.GISTIC2.R]: parameter is writeLog, default TRUE. This was announced on the Riot chat.

File: R/importCNA.GISTIC2.R [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+        write("Files exist. Now parsing.",file="test.log",append=TRUE)
+    }
+
+    # read and write files in vector
+
+
+    for (f in 1:numFiles){
+        #read file into tab delimited table
+        temp <- read.delim(fFHG[f])
+        # write to the proper place
```
Comment (hyginn) [inline R/importCNA.GISTIC2.R]: use `readr::read_delim()`

File: R/importCNA.GISTIC2.R (hunk 2 above)
Comment (nainasingh) [inline R/importCNA.GISTIC2.R]: I wrote a mechanism for creating unique filenames already. Should I change the function still?

Comment (hyginn) [inline R/importCNA.GISTIC2.R]: Yes. This really would be scope creep for the function. (And that's why we go through design -> writing tests -> implementing code in that order.)

File: R/importCNA.GISTIC2.R [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    # check if file exists
+
+    if(!file.exists(fFHG)){
+            stop(fFHG)
+    }
+
+    # parameter check
+
+    cR <- character()
+    cR <- c(cR, .checkArgs(fFHG,        like = "devCNA.txt",   checkSize = TRUE))
```
Comment (hyginn) [inline R/importCNA.GISTIC2.R]: Just a note: "like" only checks if the mode/type/class/ of the argument is the same. It does not check e.g. the actual structure of the string. Just using "a" has he same effect.

File: R/importCNA.GISTIC2.R [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+    cR <- c(cR, .checkArgs(dCNA,        like = "inst/extdata/dCNA",
```
Comment (hyginn) [inline R/importCNA.GISTIC2.R]: Just a note: `"like = ..."` only checks if the mode/type/class/ of the argument is the same. It does not check e.g. the actual structure of the string. Just using `"a"` has he same effect.

File: R/importCNA.GISTIC2.R [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+        stop(cR)
+    }
+
+    # read and write files in vector
+    # commented out part that would handle multiple files at one time
+
+    #for (f in 1:numFiles){
+        #read file into tab delimited table
+        #temp <- read.delim(fFHG)
+        temp <- readr::read_delim(file = fFHG,
```
Comment (hyginn) [inline R/importCNA.GISTIC2.R]: How will you know that this is a valid file?

File: R/importCNA.GISTIC2.R [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+                                  delim = "\t", col_names =TRUE,
+                                  col_types = c(col_character(),col_integer(),
+                                                col_character(), col_double()))
```
Comment (hyginn) [inline R/importCNA.GISTIC2.R]: How many columns will there be?

File: R/importCNA.GISTIC2.R [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+        #if (colnames(temp)[1] != "Gene Symbol"){
+            #stop("Not a valid file")
+        #}
```
Comment (hyginn) [inline R/importCNA.GISTIC2.R]: Is this check not necessary?

File: R/importCNA.GISTIC2.R [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+        # write to the proper place so it doesn't get overwritten
+        namefile<- strsplit(fFHG, "/")[[1]][3]
```
Comment (hyginn) [inline R/importCNA.GISTIC2.R]: When do you check that the values conform to the rCNA spec?

File: R/importCNA.GISTIC2.R [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+        fi <- paste0(dCNA, "/", namefile, ".rds")
```
Comment (hyginn) [inline R/importCNA.GISTIC2.R]: This is not safe for windows. Use `basename()` - also you are hardcoding this as the third element in a path that could be anything!

File: R/importCNA.GISTIC2.R [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+
+        # get rid of unwanted columns
+        #fi$Locus.ID <- NULL
+        #fi$Cytoband <- NULL
+
```
Comment (hyginn) [inline R/importCNA.GISTIC2.R]: Missing: attach metadata (cf. `.df2gG()`).

File: R/importCNA.GISTIC2.R [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+    # write to console
+
+    if (silent){
+        sprintf("All files in fFHG written into %s. Success! Import from FIREHOSE complete", dCNA)
+    }
+
+    # write to log
+
+    if(writeLog) {
+        msg    <- sprintf("event > %s from importCNA.GISTIC2()\n", Sys.time())
```
Comment (hyginn) [inline R/importCNA.GISTIC2.R]: Use the new log message structure (cf. `importNet.String()`)

File: R/importCNA.GISTIC2.R [hunk 17]
```diff
@@ -0,0 +1,96 @@
+# importCNA.GISTIC2.R
```
Comment (hyginn) [inline R/importCNA.GISTIC2.R]: I'm entering some comments here - of course coding this is secondary to coding the tests. Just some stuff I noticed.

File: tests/testthat/testImportCNA.GISTIC2.R [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+library(testthat)
+context( "import rCNA data from GISTIC2 files")
+#
```
Comment (hyginn) [inline tests/testthat/testImportCNA.GISTIC2.R]: It would be better to *design* the tests before writing them.

File: R/importCNA.GISTIC2.R (hunk 11 above)
Comment (nainasingh) [inline R/importCNA.GISTIC2.R]: This code was able to create the dataframe with several sample columns, for me, even though col_double() was only specified once

File: tests/testthat/testImportCNA.GISTIC2.R [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+#
+
```
Comment (hyginn) [inline tests/testthat/testImportCNA.GISTIC2.R]: First set of tests: invalid input is rejected.

`test_that( "importCNA.GISTIC2 rejects invalid input ...")
# missing filename
# invalid filename
# file does not exist
# missing target directory
# invalid target directory
# directory does not exist
`

File: tests/testthat/testImportCNA.GISTIC2.R [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+})
+
+test_that("bad input", {
+    expect_error(
+        importCNA.GISTIC2("tests/testthat/importhistictest.txt","inst/extdata/dCNA",
+                          FALSE, FALSE)
+    )
+})
+
+
```
Comment (hyginn) [inline tests/testthat/testImportCNA.GISTIC2.R]:
```
test_that( "valid input produces valid output") ...
# check contents
# check rownames, colnames
# check metadata
# check log-message
```

File: R/importCNA.GISTIC2.R (hunk 15 above)
Comment (nainasingh) [inline R/importCNA.GISTIC2.R]: What would be the arguments parameter in this case? 

`.df2gG <- function(inFile, call, isDirected = TRUE, simplify = TRUE)`

File: tests/testthat/testImportCNA.GISTIC2.R [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+test_that("It works", {
+    importCNA.GISTIC2("inst/extdata/devCNA.txt","inst/extdata/dCNA")
+    expect_equal(readRDS("inst/extdata/dCNA/rCNA1.rds"),
```
Comment (hyginn) [inline tests/testthat/testImportCNA.GISTIC2.R]: put test input in the `test/testthat` directory

File: tests/testthat/testImportCNA.GISTIC2.R [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+        readRDS('tests/testthat/dCNA/testimport.rds'))
+})
+
+
```
Comment (hyginn) [inline tests/testthat/testImportCNA.GISTIC2.R]:
```
test_that( "invalid input does not produce invalid output") ...
# check "NA" "na" "N/A", "", or other characters in input
# check missing line-break in last line (should this be an error condition - "incomplete file"?)
# check value outside expected range
```

File: R/importCNA.GISTIC2.R (hunk 12 above)
Comment (nainasingh) [inline R/importCNA.GISTIC2.R]: It's still "Gene Symbol" for me when I test the temp df on my console

File: R/importCNA.GISTIC2.R (hunk 15 above)
Comment (hyginn) [inline R/importCNA.GISTIC2.R]: You have an outdated version of the repository. Pull first.

File: R/importCNA.GISTIC2.R (hunk 11 above)
Comment (hyginn) [inline R/importCNA.GISTIC2.R]: Hm. Is this safe? Documented? Seems treacherous to me. Can this be written more explicitly? Or at least commented?

**Comment by hyginn** [comment]
Can we close this pull request now? You can still access comments in via the link to closed requests.
